sys/include/wtimer.h

```diff
@@ -0,0 +1,330 @@
+} wtimer_t;
+
+/**
+ * @brief wtimer initialization function
+ *
+ * This sets up wtimer. Has to be called once at system boot.
+ * If auto_init is enabled, it will call this for you.
+ */
+void wtimer_init(void);
+
+/**
+ * @brief Stop execution of a thread for some time
+ *
+ * When called from an ISR, this function will spin and thus block the MCU in
+ * interrupt context for the specified amount in *seconds*, so don't *ever* use
+ * it there,
+ *
+ * @param in seconds the amount of seconds the thread should sleep
+ */
+static void wtimer_sleep(uint32_t seconds);
+
+/**
+ * @brief Stop execution of a thread for some time
+ *
+ * When called from an ISR, this function will spin and thus block the MCU for
+ * the specified amount in microseconds, so only use it there for *very* short
+ * periods, e.g., less then WTIMER_BACKOFF.
+ *
+ * @param in microseconds the amount of microseconds the thread should sleep
+ */
+static void wtimer_usleep(uint32_t microseconds);
+
+/**
+ * @brief Stop execution of a thread for some time, 64bit version
+ *
+ * When called from an ISR, this function will spin and thus block the MCU for
+ * the specified amount in microseconds, so only use it there for *very* short
+ * periods, e.g., less then WTIMER_BACKOFF.
+ *
+ * @param in microseconds the amount of microseconds the thread should sleep
+ */
+static inline void wtimer_usleep64(uint64_t microseconds);
+
+/**
+ * @brief Stop execution of a thread for some time
+ *
+ * Don't expect nanosecond accuracy. As of mid 2015, this function just calls
+ * wtimer_usleep(nanoseconds/1000).
+ *
+ * When called from an ISR, this function will spin-block, so only use it there for
+ * *very* short periods.
+ *
+ * @param in seconds the amount of nanoseconds the thread should sleep
+ */
+static void wtimer_nanosleep(uint32_t nanoseconds);
+
+/**
+ * @brief Stop execution of a thread for some time, blocking
+ *
+ * This function will spin-block, so only use it *very* short periods.
+ *
+ * @param[in] microseconds the amount of microseconds the thread should spin
+ */
+static inline void wtimer_spin(uint32_t microseconds);
+
+ /**
+ * @brief will cause the calling thread to be suspended until the absolute
+ * time (last_wakeup + interval). last_wakeup is set to wtimer_now().
+ *
+ * This function can be used to create periodic wakeups.
+ * @c last_wakeup should be set to wtimer_now() before first call of the function.
+ *
+ * If the result of (@c last_wakeup) would be in the past, the function sets
+ * @c last_wakeup to wtimer_now() and returns immediately.
+ *
+ * @param[in] last_wakeup base time for the wakeup
+ * @param[in] usecs time in microseconds that will be added to last_wakeup
+ * @return 0 on success, < 0 on error
+ */
+int wtimer_usleep_until(uint32_t *last_wakeup, uint32_t usecs);
+
+/**
+ * @brief get the current system time as 64bit microsecond value
+ *
+ * @return current time as 64bit microsecond value
+ */
+uint64_t wtimer_now64(void);
```

Comment:
Are all platforms able to handle uint64_t return values? I'm thinking of avr8 and msp430. I know the Cortex M platforms are fine with it.